### PR TITLE
Separate GetDenominationFrequencies

### PR DESCRIPTION
**Pull request by molnard**
Breaking down https://github.com/nopara73/Sake/pull/13

### CODE REVIEW

**Review by nopara73 — CHANGES_REQUESTED** [review]
I think this prevents creating statistics from hundreds of rounds. It makes Sake slow and therefore unusable when you want to do statistically significant measurements, rather than just to get a hunch on your quick changes.

**Comment by molnard** [comment]
Every new round will recreate the denominations. That is the price we need to pay for taproot randomization. However clever things can be done but it is too early to further complicate this with early optimization. 

> slow and therefore unusable when you want to do statistically significant measurements

Give me a code that is slow and I will optimize it.

**Comment by molnard** [comment]
We need to give out a release candidate on Friday which means we have one day to merge taproot randomization in the master. How can we speed up this?

**Comment by nopara73** [comment]
> Every new round will recreate the denominations.

That's good. Then you need to only create the denominations once per round instead of once every user.

> We need to give out a release candidate on Friday

Let's delay that.

> How can we speed up this?

https://calendly.com/nopara73

**Comment by molnard** [comment]
> Let's delay that.

That is not possible. Trezor will release and an agreement was already made when I joined into the discussions. Anyway, talk to Lucas and Max regarding delaying.

**Comment by molnard** [comment]
> create the denominations once per round instead of once every user

Until I do not see how you are creating statistics from hundreds of rounds it is hard to know. Can you share your method?

**Comment by molnard** [comment]
> Then you need to only create the denominations once per round instead of once every user.

As far as I can see it is already like that. denominations generated in ctor, like before.

**Comment by molnard** [comment]
I move all the calculations that can be done in advance.

File: Sake/Mixer.cs [hunk 1]
```diff
@@ -200,40 +194,16 @@ public IEnumerable<IEnumerable<ulong>> CompleteMix(IEnumerable<IEnumerable<ulong
                     }
                 }
 
-                yield return Decompose(currentUser, others, maxVsizeCredentialValue);
+                yield return Decompose(currentUser, filteredDenominations, maxVsizeCredentialValue);
```
Comment (molnard) [inline Sake/Mixer.cs:197]: othersInputsParam is not used anymore. Passing in the denominations.

File: Sake/Mixer.cs [hunk 2]
```diff
@@ -200,40 +194,16 @@ public IEnumerable<IEnumerable<ulong>> CompleteMix(IEnumerable<IEnumerable<ulong
...
             }
         }
 
         /// <param name="maxVsizeCredentialValue">Maximum usable Vsize that client can get per alice.</param>
-        public IEnumerable<ulong> Decompose(IEnumerable<ulong> myInputsParam, IEnumerable<ulong> othersInputsParam, int maxVsizeCredentialValue)
+        public IEnumerable<ulong> Decompose(IEnumerable<ulong> myInputsParam, IEnumerable<ulong> denoms, int maxVsizeCredentialValue)
         {
-            // Filter out and order denominations those have occured in the frequency table at least twice.
-            var preFilteredDenoms = DenominationFrequencies
```
Comment (molnard) [inline Sake/Mixer.cs:211]: Moved all the calculations as is into GetFilteredDenominations.

**Comment by nopara73** [comment]
> Until I do not see how you are creating statistics from hundreds of rounds it is hard to know. Can you share your method?

```cs
for (int i = 0; i < 1_000; i++)
{
   Program.cs
}
```

**Comment by nopara73** [comment]
LGTM, even better than the original if I'm not missing something.

**Comment by nopara73** [comment]
Actually based on quick measurements it does not make a difference.

**Pull request merged**